### PR TITLE
Ben/tensorboard support

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

File: finetune/model.py [hunk 1]
```diff
@@ -136,14 +149,40 @@ def _finetune(self, *Xs, Y, batch_size=BATCH_SIZE):
         self._build_model(n_updates_total=n_updates_total, n_classes=self.n_classes)
 
         dataset = shuffle(train_x, train_mask, Y, random_state=np.random)
+        x_t, x_v, m_t, m_v, y_t, y_v = train_test_split(*dataset, test_size=val_frac, random_state=31415)
```
Comment (benleetownsend) [inline finetune/model.py]: change to tr and va

File: finetune/model.py [hunk 2]
```diff
@@ -136,14 +149,40 @@ def _finetune(self, *Xs, Y, batch_size=BATCH_SIZE):
...
+        best_val_loss = float("inf")
         for i in range(N_EPOCHS):
             for xmb, mmb, ymb in iter_data(*dataset, n_batch=n_batch_train, verbose=True):
-                cost, _ = self.sess.run([self.clf_loss, self.train_op], {self.X: xmb, self.M: mmb, self.Y: ymb})
-                rolling_avg_loss = rolling_avg_loss * 0.95 + cost *  0.05
-                print("\nLOSS = {}, ROLLING AVG = {}".format(cost, rolling_avg_loss))
+                global_step += 1
+                if global_step % eval_interval == 0:
+                    sum_val_loss = 0
+                    for val_step, (xval, mval, yval) in enumerate(
```
Comment (benleetownsend) [inline finetune/model.py]: remove enumerate

File: finetune/model.py [hunk 3]
```diff
@@ -136,14 +149,39 @@ def _finetune(self, *Xs, Y, batch_size=BATCH_SIZE):
...
+                global_step += 1
+                if global_step % eval_interval == 0:
+                    sum_val_loss = 0
+                    for xval, mval, yval in enumerate(iter_data(*val_dataset, n_batch=n_batch_train, verbose=True)):
+                        val_cost, summary = self.sess.run([self.clf_loss, self.summaries],
+                                                          {self.X: xval, self.M: mval, self.Y: yval})
+                        self.valid_writer.add_summary(summary, global_step)
+                        sum_val_loss += val_cost
+                        avg_val_loss = avg_val_loss * ROLLING_AVG_DECAY + val_cost * (1 - ROLLING_AVG_DECAY)
+                        print("\nVAL: LOSS = {}, ROLLING AVG = {}".format(val_cost, avg_val_loss))
```
Comment (madisonmay) [inline finetune/model.py]: Would recommend either using the verbose parameter to decide whether to write to stdout or use the logging module with a logging.INFO level so folks can opt out of getting stdout output.

File: finetune/model.py [hunk 4]
```diff
@@ -356,7 +408,9 @@ def _load_base_model(self):
         init_params = np.split(np.concatenate(init_params, 0), offsets)[:-1]
         init_params = [param.reshape(shape) for param, shape in zip(init_params, shapes)]
         init_params[0] = init_params[0][:self.max_length]
-        init_params[0] = np.concatenate([init_params[1], (np.random.randn(len(self.encoder.special_tokens), N_EMBED) * WEIGHT_STDDEV).astype(np.float32), init_params[0]], 0)
+        init_params[0] = np.concatenate([init_params[1], (
+                np.random.randn(len(self.encoder.special_tokens), N_EMBED) * WEIGHT_STDDEV).astype(np.float32),
```
Comment (madisonmay) [inline finetune/model.py]: assign to var to make a bit more readable

File: finetune/model.py [hunk 5]
```diff
@@ -521,9 +575,11 @@ def featurize(self, X_1, X_2, max_length=None):
         """
         return self._featurize(X_1, X_2, max_length=max_length)
 
+
 if __name__ == "__main__":
 
     import json
```
Comment (madisonmay) [inline finetune/model.py]: move to top

**Comment by madisonmay** [comment]
Few minor comments but otherwise LGTM.  Just merge yourself once you make those edits + get passing tests.

File: finetune/model.py [hunk 6]
```diff
@@ -123,7 +136,7 @@ def _text_to_ids(self, *Xs, max_length=None):
         tokens, mask = self._array_format(token_idxs)
         return tokens, mask
 
-    def _finetune(self, *Xs, Y, batch_size=BATCH_SIZE):
+    def _finetune(self, *Xs, Y, batch_size=BATCH_SIZE, val_frac=0.05, eval_interval=150):
```
Comment (madisonmay) [inline finetune/model.py]: rename to validation_size and allow int / float (we get this for free thanks to `train_test_split` being used)

**Pull request merged**